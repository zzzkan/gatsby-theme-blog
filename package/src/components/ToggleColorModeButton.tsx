import React from "react"
import { IconButton, useColorMode } from "@chakra-ui/react"
import { RiSunLine } from "@react-icons/all-files/ri/RiSunLine"
import { RiMoonLine } from "@react-icons/all-files/ri/RiMoonLine"
import { useMultilingualSentence } from "../hooks/useMultilingualSentence"

export const ToggleColorModeButton: React.FC = () => {
  const { colorMode, toggleColorMode } = useColorMode()
  const { getToggleColorModeLabel } = useMultilingualSentence()
  return (
    <IconButton
      onClick={toggleColorMode}
      aria-label={getToggleColorModeLabel()}
      variant={"ghost"}
      isRound={true}
      size={"sm"}
      fontSize={"2xl"}
      icon={colorMode === "light" ? <RiMoonLine /> : <RiSunLine />}
    />
  )
}
