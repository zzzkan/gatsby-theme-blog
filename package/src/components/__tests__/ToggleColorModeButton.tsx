import React from "react"
import { render } from "@testing-library/react"
import { ToggleColorModeButton } from "../ToggleColorModeButton"
import * as chakra from "@chakra-ui/react"
import { RiSunLine } from "@react-icons/all-files/ri/RiSunLine"
import { RiMoonLine } from "@react-icons/all-files/ri/RiMoonLine"

describe("ToggleColorModeButton component", () => {
  vi.mock("@react-icons/all-files/ri/RiSunLine")
  vi.mock("@react-icons/all-files/ri/RiMoonLine")
  const mockRiSunLine = RiSunLine as jest.Mock
  const mockRiMoonLine = RiMoonLine as jest.Mock

  afterEach(() => {
    vi.clearAllMocks()
  })

  afterAll(() => {
    vi.restoreAllMocks()
  })

  test("snapshot", () => {
    const { asFragment } = render(<ToggleColorModeButton />)
    expect(asFragment()).toMatchSnapshot()
  })

  test("should render light icon", () => {
    vi.spyOn(chakra, "useColorMode").mockImplementationOnce(() => ({
      colorMode: "light",
      toggleColorMode: vi.fn(),
      setColorMode: vi.fn(),
    }))
    render(<ToggleColorModeButton />)
    expect(mockRiSunLine).not.toHaveBeenCalled()
    expect(mockRiMoonLine).toHaveBeenCalled()
  })

  test("should render dark icon", () => {
    vi.spyOn(chakra, "useColorMode").mockImplementationOnce(() => ({
      colorMode: "dark",
      toggleColorMode: vi.fn(),
      setColorMode: vi.fn(),
    }))
    render(<ToggleColorModeButton />)
    expect(mockRiSunLine).toHaveBeenCalled()
    expect(mockRiMoonLine).not.toHaveBeenCalled()
  })
})
