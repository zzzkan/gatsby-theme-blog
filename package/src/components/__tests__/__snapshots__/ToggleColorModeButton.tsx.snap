// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ToggleColorModeButton component > snapshot 1`] = `
<DocumentFragment>
  .emotion-0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  position: relative;
  white-space: nowrap;
  vertical-align: middle;
  outline: 2px solid transparent;
  outline-offset: 2px;
  padding: 0px;
  border-radius: full;
  font-size: 2xl;
}

<button
    aria-label="Switch to color mode"
    class="chakra-button emotion-0"
    type="button"
  />
</DocumentFragment>
`;
